File: packages/webreel/src/lib/runner.ts
import { resolve, dirname } from "node:path";
import { readFileSync, writeFileSync, mkdirSync, rmSync, existsSync } from "node:fs";
import { pathToFileURL } from "node:url";
import {
  type CDPClient,
  type BoundingBox,
  type OverlayTheme,
  type AutoZoomConfig,
  type ZoomEvent,
  RecordingContext,
  connectCDP,
  launchChrome,
  navigate,
  waitForSelector,
  waitForText,
  injectOverlays,
  pause,
  findElementByText,
  findElementBySelector,
  clickAt,
  pressKey,
  typeText,
  dragFromTo,
  moveCursorTo,
  captureScreenshot,
  Recorder,
  InteractionTimeline,
  compose,
  ensureFfmpeg,
  extractThumbnail,
  moveFileSync,
  buildAutoZoomFilter,
  unionBboxes,
  installRevealObserver,
  collectReveals,
  type RevealObserverHandle,
  DEFAULT_VIEWPORT_SIZE,
} from "@webreel/core";
import type { VideoConfig, Step, ElementTarget } from "./types.js";
import { registerInterruptCleanup } from "./signals.js";

export function formatStep(i: number, step: Step): string {
  const desc = "description" in step && step.description ? `: ${step.description}` : "";
  switch (step.action) {
    case "pause":
      return `[step ${i}] pause ${step.ms}ms${desc}`;
    case "click":
      return `[step ${i}] click ${step.text ? `text="${step.text}"` : `selector="${step.selector}"`}${desc}`;
    case "key":
      return `[step ${i}] key "${step.key}"${desc}`;
    case "type":
      return `[step ${i}] type "${step.text}"${desc}`;
    case "scroll":
      return `[step ${i}] scroll x=${step.x ?? 0} y=${step.y ?? 0}${desc}`;
    case "wait":
      return `[step ${i}] wait ${step.selector ? `selector="${step.selector}"` : `text="${step.text}"`}${desc}`;
    case "drag":
      return `[step ${i}] drag${desc}`;
    case "moveTo":
      return `[step ${i}] moveTo ${step.text ? `text="${step.text}"` : `selector="${step.selector}"`}${desc}`;
    case "screenshot":
      return `[step ${i}] screenshot "${step.output}"${desc}`;
    case "navigate":
      return `[step ${i}] navigate "${step.url}"${desc}`;
    case "navigateHref":
      return `[step ${i}] navigateHref selector="${step.selector}"${desc}`;
    case "hover":
      return `[step ${i}] hover ${step.text ? `text="${step.text}"` : `selector="${step.selector}"`}${desc}`;
    case "select":
      return `[step ${i}] select "${step.selector}" value="${step.value}"${desc}`;
    case "upload":
      return `[step ${i}] upload selector="${step.selector}" file="${step.filePath}"${desc}`;
    default: {
      const _exhaustive: never = step;
      return `[step ${i}] ${(_exhaustive as Step).action}`;
    }
  }
}

export function resolveKeyTarget(target: string | ElementTarget): string {
  if (typeof target === "string") return target;
  return target.selector ?? "";
}

export function resolveTypeMethod(step: {
  method?: "insertText" | "dispatchKeyEvent";
  selector?: string;
}): "insertText" | "dispatchKeyEvent" {
  return step.method ?? (step.selector ? "insertText" : "dispatchKeyEvent");
}

export async function resolveTarget(
  client: CDPClient,
  opts: { text?: string; selector?: string; within?: string },
): Promise<BoundingBox> {
  if (!opts.text && !opts.selector) {
    throw new Error(`resolveTarget requires "text" or "selector"`);
  }
  let box: BoundingBox | null = null;
  if (opts.text) {
    box = await findElementByText(client, opts.text, opts.within);
  } else if (opts.selector) {
    box = await findElementBySelector(client, opts.selector, opts.within);
  }
  if (!box) {
    const target = opts.text ? `text="${opts.text}"` : `selector="${opts.selector}"`;
    const scope = opts.within ? ` within "${opts.within}"` : "";
    throw new Error(`Element not found: ${target}${scope}`);
  }
  return box;
}

export function resolveUrl(url: string, baseUrl: string, configDir: string): string {
  if (url.startsWith("http") || url.startsWith("file://")) return url;
  const combined = `${baseUrl}${url}`;
  if (combined.startsWith("http") || combined.startsWith("file://")) return combined;
  return pathToFileURL(resolve(configDir, combined)).href;
}

export function randomPointInBox(
  box: BoundingBox,
  spread = 0.25,
): { x: number; y: number } {
  const center = 0.5 - spread / 2;
  return {
    x: box.x + box.width * (center + Math.random() * spread),
    y: box.y + box.height * (center + Math.random() * spread),
  };
}

export function normalizeAutoZoom(value: VideoConfig["autoZoom"]): AutoZoomConfig {
  if (value === true) return { enabled: true };
  if (value === false || value === undefined) return { enabled: false };
  return { ...value, enabled: value.enabled ?? true };
}

async function captureZoomEvent(
  client: CDPClient,
  step: Step,
  timeline: InteractionTimeline,
  preEventCount: number,
  fps: number,
  previousBox: BoundingBox | null,
  reveals: BoundingBox[] = [],
): Promise<ZoomEvent | null> {
  let box: BoundingBox | null = null;

  switch (step.action) {
    case "navigate":
    case "navigateHref":
    case "pause":
    case "wait":
    case "screenshot":
    case "scroll":
    case "upload":
      return null;

    case "click":
    case "moveTo":
    case "hover":
    case "select": {
      if (step.selector) {
        box = await findElementBySelector(client, step.selector, step.within);
      } else if (step.text) {
        box = await findElementByText(client, step.text, step.within);
      }
      break;
    }

    case "type": {
      if (step.selector) {
        box = await findElementBySelector(client, step.selector, step.within);
      } else if (previousBox) {
        box = previousBox;
      }
      break;
    }

    case "key": {
      if (step.target) {
        const sel = typeof step.target === "string" ? step.target : step.target.selector;
        const within = typeof step.target === "string" ? undefined : step.target.within;
        if (sel) box = await findElementBySelector(client, sel, within);
      } else if (previousBox) {
        box = previousBox;
      }
      break;
    }

    case "drag": {
      if (step.from.selector) {
        box = await findElementBySelector(client, step.from.selector, step.from.within);
      } else if (step.from.text) {
        box = await findElementByText(client, step.from.text, step.from.within);
      }
      break;
    }
  }

  if (!box) return null;

  // For click/drag, union the click target's bbox with any elements that
  // newly appeared or grew in size during the step (dropdowns, modals,
  // tooltips). This makes the zoom frame the whole widget rather than just
  // the trigger, matching Cursor's behavior for button-to-modal interactions.
  if ((step.action === "click" || step.action === "drag") && reveals.length > 0) {
    const unioned = unionBboxes([box, ...reveals]);
    if (unioned) box = unioned;
  }

  // Pick the timestamp from SoundEvents that fired DURING this step so the
  // zoom anticipation is anchored on the actual interaction moment (not the
  // step's end-of-postDelay time, which lags the click by 1-2 seconds for
  // steps with cursor animation).
  const newEvents = timeline.getEvents().slice(preEventCount);
  const fallbackMs = (timeline.getFrameCount() / fps) * 1000;
  let timeMs: number;
  let holdUntilMs: number | undefined;
  if (step.action === "click" || step.action === "drag") {
    const firstClick = newEvents.find((e) => e.type === "click");
    timeMs = firstClick ? firstClick.timeMs : fallbackMs;
  } else if (step.action === "type" || step.action === "key") {
    // Anchor the approach on the FIRST event (click that lands on the input)
    // so the camera is already zoomed in when typing begins. `holdUntilMs`
    // extends the hold through the LAST keystroke, so the camera stays on
    // the field for the entire typing span.
    const firstEvent = newEvents[0];
    const lastEvent = newEvents[newEvents.length - 1];
    timeMs = firstEvent ? firstEvent.timeMs : fallbackMs;
    if (lastEvent && lastEvent !== firstEvent) {
      holdUntilMs = lastEvent.timeMs;
    }
  } else {
    timeMs = fallbackMs;
  }

  const { result } = await client.Runtime.evaluate({
    expression: "location.href",
    returnByValue: true,
  });
  const url = typeof result.value === "string" ? result.value : undefined;
  return { timeMs, box, url, holdUntilMs };
}

export async function extractThumbnailIfConfigured(
  config: Pick<VideoConfig, "thumbnail">,
  outputPath: string,
): Promise<void> {
  if (config.thumbnail?.enabled === false) return;
  const thumbTime = config.thumbnail?.time ?? 0;
  const thumbPath = outputPath.replace(/\.[^.]+$/, ".png");
  const ffmpegPath = await ensureFfmpeg();
  extractThumbnail(ffmpegPath, outputPath, thumbPath, thumbTime);
  console.log(`Thumbnail: ${thumbPath}`);
}

export interface RunVideoOptions {
  record?: boolean;
  verbose?: boolean;
  configDir?: string;
  frames?: boolean;
}

export async function runVideo(
  config: VideoConfig,
  options?: RunVideoOptions,
): Promise<void> {
  const shouldRecord = options?.record ?? true;
  const verbose = options?.verbose ?? false;
  const saveFrames = options?.frames ?? false;
  const configDir = options?.configDir ?? process.cwd();

  console.log(`${shouldRecord ? "Recording" : "Previewing"}: ${config.name}`);

  const width = config.viewport?.width ?? DEFAULT_VIEWPORT_SIZE;
  const height = config.viewport?.height ?? DEFAULT_VIEWPORT_SIZE;
  const zoom = config.zoom ?? 1;
  const cssWidth = Math.round(width / zoom);
  const cssHeight = Math.round(height / zoom);

  const ctx = new RecordingContext();
  ctx.resetCursorPosition(cssWidth, cssHeight);
  if (config.clickDwell !== undefined) ctx.setClickDwell(config.clickDwell);
  const initialCursor = ctx.getCursorPosition();

  const chrome = await launchChrome({ headless: shouldRecord });
  let clientRef: CDPClient | null = null;
  let recorder: Recorder | null = null;

  // If the process is interrupted, release everything this run owns:
  // flush and stop ffmpeg, discard the partial capture, close the CDP
  // connection, and kill Chrome (which also removes its temp profile).
  const unregisterCleanup = registerInterruptCleanup(async () => {
    if (recorder) {
      const tempVideo = recorder.getTempVideoPath();
      try {
        await recorder.stop();
      } catch {
        // Best-effort shutdown; Chrome teardown below still runs.
      }
      rmSync(tempVideo, { force: true });
    }
    if (clientRef) {
      try {
        await clientRef.close();
      } catch {
        // Connection may already be gone.
      }
    }
    await chrome.kill();
  });

  let connectionLost: Error | null = null;

  try {
    const client = await connectCDP(chrome.port, (err) => {
      connectionLost = err;
      console.error(err.message);
    });
    clientRef = client;
    await client.Page.enable();
    await client.Runtime.enable();
    await client.Emulation.setDeviceMetricsOverride({
      width: cssWidth,
      height: cssHeight,
      deviceScaleFactor: zoom,
      mobile: false,
    });

    const baseUrl = config.baseUrl ?? "";
    const url = resolveUrl(config.url, baseUrl, configDir);

    await navigate(client, url);

    if (config.waitFor) {
      if (typeof config.waitFor === "string") {
        await waitForSelector(client, config.waitFor);
      } else if (config.waitFor.selector) {
        await waitForSelector(client, config.waitFor.selector);
      } else if (config.waitFor.text) {
        await waitForText(client, config.waitFor.text, config.waitFor.within);
      }
    }

    await pause(200);

    let overlayTheme: OverlayTheme | undefined;
    let cursorSvg: string | undefined;
    const cursorConfig = config.theme?.cursor;
    if (config.theme) {
      overlayTheme = {
        cursorSize: cursorConfig?.size,
        cursorHotspot: cursorConfig?.hotspot,
        hud: config.theme.hud,
      };
      if (cursorConfig?.image) {
        try {
          cursorSvg = readFileSync(resolve(configDir, cursorConfig.image), "utf-8");
          overlayTheme.cursorSvg = cursorSvg;
        } catch {
          throw new Error(
            `Failed to read cursor SVG: ${resolve(configDir, cursorConfig.image)}`,
          );
        }
      }
    }

    let timeline: InteractionTimeline | null = null;
    const autoZoomCfg = normalizeAutoZoom(config.autoZoom);
    const zoomEvents: ZoomEvent[] = [];
    const outputPath =
      config.output ?? resolve(configDir, "videos", `${config.name}.mp4`);

    if (shouldRecord) {
      ctx.setMode("record");
      timeline = new InteractionTimeline(width, height, {
        zoom,
        fps: config.fps,
        initialCursor,
        cursorSvg,
        cursorSize: cursorConfig?.size,
        cursorHotspot: cursorConfig?.hotspot,
        hud: config.theme?.hud,
      });
      ctx.setTimeline(timeline);

      const crf =
        config.quality !== undefined
          ? Math.round(51 * (1 - config.quality / 100))
          : undefined;
      const framesDir = saveFrames
        ? resolve(configDir, ".webreel", "frames", config.name)
        : undefined;
      recorder = new Recorder(width, height, {
        fps: config.fps,
        crf,
        framesDir,
        sfx: config.sfx,
      });
      recorder.setTimeline(timeline);
      await recorder.start(client, outputPath, ctx);
    } else {
      ctx.setMode("preview");
      ctx.setTimeline(null);
      await injectOverlays(client, overlayTheme, initialCursor);
    }

    await pause(500);

    for (let i = 0; i < config.steps.length; i++) {
      const step = config.steps[i];
      if (verbose) console.log(formatStep(i, step));
      const stepStartMs = timeline?.getCurrentTimeMs();

      let preEventCount = 0;
      let revealHandle: RevealObserverHandle | null = null;
      if (shouldRecord && autoZoomCfg.enabled && timeline) {
        preEventCount = timeline.getEvents().length;
        // Observer installed only for interactions that can reveal UI.
        // Collected after postDelay so any animations have had time to
        // settle.
        if (step.action === "click" || step.action === "drag") {
          revealHandle = await installRevealObserver(client);
        }
      }

      try {
        switch (step.action) {
          case "pause":
            await pause(step.ms);
            break;

          case "click": {
            const box = await resolveTarget(client, step);
            const { x: cx, y: cy } = randomPointInBox(box);
            await clickAt(ctx, client, cx, cy, step.modifiers);
            break;
          }

          case "key": {
            if (step.target) {
              const sel = resolveKeyTarget(step.target);
              if (sel) {
                await client.Runtime.evaluate({
                  expression: `document.querySelector(${JSON.stringify(sel)})?.focus()`,
                });
                await pause(100);
              }
            }
            await pressKey(ctx, client, step.key, step.label);
            break;
          }

          case "drag": {
            const fromBox = await resolveTarget(client, step.from);
            const toBox = await resolveTarget(client, step.to);
            await dragFromTo(ctx, client, fromBox, toBox);
            break;
          }

          case "type": {
            if (step.selector) {
              const box = await resolveTarget(client, {
                selector: step.selector,
                within: step.within,
              });
              const { x: tx, y: ty } = randomPointInBox(box);
              await clickAt(ctx, client, tx, ty);
              await client.Runtime.evaluate({
                expression: `document.querySelector(${JSON.stringify(step.selector)})?.focus()`,
              });
              await pause(300 + Math.random() * 200);
            }
            await typeText(ctx, client, step.text, step.charDelay, {
              method: resolveTypeMethod(step),
            });
            break;
          }

          case "scroll": {
            const scrollX = step.x ?? 0;
            const scrollY = step.y ?? 0;
            if (step.selector) {
              const withinScope = step.within
                ? `document.querySelector(${JSON.stringify(step.within)})`
                : "document";
              await client.Runtime.evaluate({
                expression: `(() => {
                  const scope = ${withinScope};
                  const target = scope?.querySelector(${JSON.stringify(step.selector)});
                  if (target) target.scrollBy({ left: ${scrollX}, top: ${scrollY}, behavior: "smooth" });
                })()`,
              });
            } else if (step.text) {
              const box = await resolveTarget(client, step);
              await client.Runtime.evaluate({
                expression: `(() => {
                  const el = document.elementFromPoint(${Math.round(box.x + box.width / 2)}, ${Math.round(box.y + box.height / 2)});
                  if (el) el.scrollBy({ left: ${scrollX}, top: ${scrollY}, behavior: "smooth" });
                })()`,
              });
            } else {
              await client.Runtime.evaluate({
                expression: `window.scrollBy({ left: ${scrollX}, top: ${scrollY}, behavior: "smooth" })`,
              });
            }
            await pause(500);
            break;
          }

          case "wait": {
            if (step.selector) {
              const scopedSelector = step.within
                ? `${step.within} ${step.selector}`
                : step.selector;
              await waitForSelector(client, scopedSelector, step.timeout);
            } else if (step.text) {
              await waitForText(client, step.text, step.within, step.timeout);
            }
            break;
          }

          case "screenshot": {
            await captureScreenshot(client, resolve(configDir, step.output));
            break;
          }

          case "moveTo": {
            const box = await resolveTarget(client, step);
            const { x: mx, y: my } = randomPointInBox(box, 0.1);
            await moveCursorTo(ctx, client, mx, my);
            break;
          }

          case "navigate": {
            const navUrl = resolveUrl(step.url, config.baseUrl ?? "", configDir);
            await navigate(client, navUrl);
            break;
          }

          case "navigateHref": {
            // Runtime.evaluate does not surface in-page exceptions, so the
            // probe returns sentinels for the two expected failure modes.
            const result = await client.Runtime.evaluate({
              expression: `(() => {
                const el = document.querySelector(${JSON.stringify(step.selector)});
                if (!el) return "__WEBREEL_NOT_FOUND__";
                const href = el.href || el.getAttribute("href");
                if (!href) return "__WEBREEL_NO_HREF__";
                return new URL(href, window.location.href).href;
              })()`,
              returnByValue: true,
            });
            const href = result.result?.value;
            if (href === "__WEBREEL_NOT_FOUND__") {
              throw new Error(`Element not found: selector="${step.selector}"`);
            }
            if (href === "__WEBREEL_NO_HREF__") {
              throw new Error(`Element has no href: selector="${step.selector}"`);
            }
            if (typeof href !== "string" || href.length === 0) {
              throw new Error("navigateHref did not resolve a URL");
            }
            await navigate(client, href);
            break;
          }

          case "hover": {
            const box = await resolveTarget(client, step);
            const { x: hx, y: hy } = randomPointInBox(box, 0.1);
            await moveCursorTo(ctx, client, hx, hy);
            await client.Input.dispatchMouseEvent({
              type: "mouseMoved",
              x: hx,
              y: hy,
            });
            break;
          }

          case "select": {
            if (step.selector) {
              const withinScope = step.within
                ? `document.querySelector(${JSON.stringify(step.within)})`
                : "document";
              await client.Runtime.evaluate({
                expression: `(() => {
                  const scope = ${withinScope};
                  const el = scope?.querySelector(${JSON.stringify(step.selector)});
                  if (!el) throw new Error("Element not found: " + ${JSON.stringify(step.selector)});
                  el.value = ${JSON.stringify(step.value)};
                  el.dispatchEvent(new Event("input", { bubbles: true }));
                  el.dispatchEvent(new Event("change", { bubbles: true }));
                })()`,
              });
            } else if (step.text) {
              const box = await resolveTarget(client, step);
              await client.Runtime.evaluate({
                expression: `(() => {
                  const el = document.elementFromPoint(${Math.round(box.x + box.width / 2)}, ${Math.round(box.y + box.height / 2)});
                  if (!el) throw new Error("Element not found by text: " + ${JSON.stringify(step.text)});
                  el.value = ${JSON.stringify(step.value)};
                  el.dispatchEvent(new Event("input", { bubbles: true }));
                  el.dispatchEvent(new Event("change", { bubbles: true }));
                })()`,
              });
            } else {
              throw new Error(`select step requires "selector" or "text"`);
            }
            break;
          }

          case "upload": {
            const absolutePath = resolve(configDir, step.filePath);
            if (!existsSync(absolutePath)) {
              throw new Error(`File not found: ${absolutePath}`);
            }
            await client.DOM.enable();
            const { root } = await client.DOM.getDocument({ depth: 0 });
            const { nodeId } = await client.DOM.querySelector({
              nodeId: root.nodeId,
              selector: step.selector,
            });
            if (!nodeId) {
              throw new Error(`File input not found: selector="${step.selector}"`);
            }
            await client.DOM.setFileInputFiles({ nodeId, files: [absolutePath] });
            break;
          }
        }
        const stepDelay = "delay" in step ? step.delay : undefined;
        const postDelay = stepDelay ?? config.defaultDelay;
        if (postDelay !== undefined && postDelay > 0) {
          await pause(postDelay);
        }
        if (timeline && stepStartMs !== undefined) {
          timeline.addStep({
            index: i,
            action: step.action,
            startMs: stepStartMs,
            endMs: timeline.getCurrentTimeMs(),
            ...(step.label ? { label: step.label } : {}),
            ...(step.description ? { description: step.description } : {}),
          });
        }
        if (shouldRecord && autoZoomCfg.enabled && timeline) {
          const fps = config.fps ?? timeline.toJSON().fps;
          const previousBox =
            zoomEvents.length > 0 ? zoomEvents[zoomEvents.length - 1].box : null;
          const reveals: BoundingBox[] = revealHandle
            ? await collectReveals(client, revealHandle)
            : [];
          if (process.env.WEBREEL_DEBUG_ZOOM && revealHandle) {
            const summary = reveals
              .map(
                (r) =>
                  `${r.x.toFixed(0)},${r.y.toFixed(0)} ${r.width.toFixed(0)}x${r.height.toFixed(0)}`,
              )
              .join(" | ");
            console.error(
              `reveals (${step.action}): ${reveals.length}${summary ? " " + summary : ""}`,
            );
          }
          const ze = await captureZoomEvent(
            client,
            step,
            timeline,
            preEventCount,
            fps,
            previousBox,
            reveals,
          );
          if (ze) zoomEvents.push(ze);
        }
      } catch (err) {
        throw new Error(
          `Step ${i} (${step.action}) failed at ${url}: ${err instanceof Error ? err.message : String(err)}`,
          { cause: err },
        );
      }

      if (connectionLost) throw connectionLost;
    }

    if (connectionLost) throw connectionLost;

    if (recorder) {
      const cleanVideoPath = recorder.getTempVideoPath();
      await recorder.stop();
      recorder = null;

      if (timeline) {
        const timelineData = timeline.toJSON();
        if (zoomEvents.length > 0) {
          timelineData.zoomEvents = zoomEvents;
        }
        const metadataDir = resolve(configDir, ".webreel", "timelines");
        mkdirSync(metadataDir, { recursive: true });
        writeFileSync(
          resolve(metadataDir, `${config.name}.timeline.json`),
          JSON.stringify(timelineData),
        );

        const rawDir = resolve(configDir, ".webreel", "raw");
        mkdirSync(rawDir, { recursive: true });
        const rawVideoPath = resolve(rawDir, `${config.name}.mp4`);
        moveFileSync(cleanVideoPath, rawVideoPath);

        ctx.setMode("preview");
        ctx.setTimeline(null);
        mkdirSync(dirname(outputPath), { recursive: true });
        console.log(`Compositing overlays...`);
        const zoomFilter = autoZoomCfg.enabled
          ? buildAutoZoomFilter(
              zoomEvents,
              { width: timelineData.width, height: timelineData.height },
              timelineData.zoom ?? 1,
              timelineData.frames.length / timelineData.fps,
              timelineData.fps,
              autoZoomCfg,
            )
          : null;
        if (zoomFilter) {
          console.log(
            `Applying autozoom (${zoomEvents.length} event${zoomEvents.length === 1 ? "" : "s"})`,
          );
          if (verbose) {
            for (const e of zoomEvents) {
              const box = e.box;
              console.log(
                `  t=${(e.timeMs / 1000).toFixed(2)}s box=${box.x.toFixed(0)},${box.y.toFixed(0)} ${box.width.toFixed(0)}x${box.height.toFixed(0)}`,
              );
            }
          }
        }
        await compose(rawVideoPath, timelineData, outputPath, {
          sfx: config.sfx,
          zoomFilter: zoomFilter ?? undefined,
        });
      }
      await extractThumbnailIfConfigured(config, outputPath);

      console.log(`Done: ${outputPath}`);
    } else {
      console.log(`Preview complete: ${config.name}`);
    }
  } finally {
    unregisterCleanup();
    if (recorder) {
      try {
        await recorder.stop();
      } catch (err) {
        console.warn("Failed to stop recorder:", err);
      }
    }
    if (clientRef) {
      try {
        await clientRef.close();
      } catch (err) {
        console.warn("Failed to close CDP client:", err);
      }
    }
    try {
      await chrome.kill();
    } catch (err) {
      console.warn("Failed to kill Chrome process:", err);
    }
  }
}

File: packages/@webreel/compositor/src/runner.zig
const std = @import("std");
const cdp = @import("cdp.zig");
const actions = @import("actions.zig");
const timeline_mod = @import("timeline.zig");
const cursor_motion = @import("cursor_motion.zig");

pub const StepAction = enum {
    pause,
    click,
    key,
    @"type",
    scroll,
    wait,
    drag,
    moveTo,
    screenshot,
    navigate,
    hover,
    select,
};

pub const StepTarget = struct {
    text: ?[]const u8 = null,
    selector: ?[]const u8 = null,
    within: ?[]const u8 = null,
};

pub const Step = struct {
    action: StepAction,
    ms: ?u64 = null,
    text: ?[]const u8 = null,
    selector: ?[]const u8 = null,
    within: ?[]const u8 = null,
    key: ?[]const u8 = null,
    label: ?[]const u8 = null,
    x: ?i32 = null,
    y: ?i32 = null,
    url: ?[]const u8 = null,
    output: ?[]const u8 = null,
    value: ?[]const u8 = null,
    delay: ?u64 = null,
    char_delay: ?u64 = null,
    timeout: ?u64 = null,
    modifiers: ?u32 = null,
    from: ?StepTarget = null,
    to: ?StepTarget = null,
    target: ?[]const u8 = null,
};

pub const VideoConfig = struct {
    name: []const u8,
    url: []const u8,
    base_url: []const u8 = "",
    output: ?[]const u8 = null,
    width: u32 = actions.DEFAULT_VIEWPORT_SIZE,
    height: u32 = actions.DEFAULT_VIEWPORT_SIZE,
    zoom: f64 = 1.0,
    fps: u32 = 60,
    crf: u32 = 18,
    color_scheme: ?[]const u8 = null,
    default_delay: ?u64 = null,
    click_dwell: ?u64 = null,
    steps: []const Step = &.{},
    ffmpeg_path: []const u8 = "ffmpeg",
    cursor_svg_path: ?[]const u8 = null,
    font_path: ?[]const u8 = null,
    screen_width: ?u32 = null,
    screen_height: ?u32 = null,
};

pub fn resolveTarget(client: *cdp.CdpClient, target: StepTarget, timeout_ms: u64) !actions.BoundingBox {
    if (target.text == null and target.selector == null) {
        return error.NoTargetSpecified;
    }

    const start = std.time.milliTimestamp();
    while (true) {
        var box: ?actions.BoundingBox = null;
        if (target.text) |text| {
            box = actions.findElementByText(client, text, target.within) catch null;
        } else if (target.selector) |sel| {
            box = actions.findElementBySelector(client, sel, target.within) catch null;
        }

        if (box) |b| {
            if (b.width > 0 and b.height > 0) return b;
        }

        if (@as(u64, @intCast(std.time.milliTimestamp() - start)) >= timeout_ms) {
            return error.ElementNotFound;
        }
        actions.sleepMs(200);
    }
}

pub fn executeStep(ctx: *actions.RecordingContext, client: *cdp.CdpClient, step: Step) !void {
    switch (step.action) {
        .pause => {
            actions.sleepMs(step.ms orelse 1200);
        },
        .click => {
            const target = StepTarget{
                .text = step.text,
                .selector = step.selector,
                .within = step.within,
            };
            const box = try resolveTarget(client, target, step.timeout orelse 10000);
            const pt = actions.randomPointInBox(box, 0.25);
            try actions.clickAt(ctx, client, pt.x, pt.y, step.modifiers);
        },
        .key => {
            const key_str = step.key orelse return error.MissingKey;
            if (step.target) |sel| {
                var expr: std.ArrayList(u8) = .empty;
                defer expr.deinit(ctx.allocator);
                var ew = expr.writer(ctx.allocator);
                try ew.writeAll("document.querySelector(\"");
                try ew.writeAll(sel);
                try ew.writeAll("\")?.focus()");
                _ = client.evaluate(expr.items) catch {};
                actions.sleepMs(100);
            }
            try actions.pressKey(ctx, client, key_str, step.label);
        },
        .@"type" => {
            const text = step.text orelse return error.MissingText;
            if (step.selector) |_| {
                const target = StepTarget{
                    .text = step.text,
                    .selector = step.selector,
                    .within = step.within,
                };
                const box = try resolveTarget(client, target, step.timeout orelse 10000);
                const pt = actions.randomPointInBox(box, 0.25);
                try actions.clickAt(ctx, client, pt.x, pt.y, null);
                actions.sleepMsF(300.0 + prng() * 200.0);
            }
            try actions.typeText(ctx, client, text, step.char_delay);
        },
        .scroll => {
            const sx = step.x orelse 0;
            const sy = step.y orelse 0;
            try actions.scroll(client, step.selector, sx, sy);
        },
        .wait => {
            const timeout = step.timeout orelse 30000;
            if (step.selector) |sel| {
                try actions.waitForSelector(client, sel, timeout);
            } else if (step.text) |text| {
                try actions.waitForText(client, text, step.within, timeout);
            }
        },
        .drag => {
            const from_target = step.from orelse return error.MissingDragFrom;
            const to_target = step.to orelse return error.MissingDragTo;
            const from_box = try resolveTarget(client, from_target, step.timeout orelse 10000);
            const to_box = try resolveTarget(client, to_target, step.timeout orelse 10000);

            const fx = from_box.x + from_box.width / 2.0;
            const fy = from_box.y + from_box.height / 2.0;
            const tx = to_box.x + to_box.width / 2.0;
            const ty = to_box.y + to_box.height / 2.0;

            try actions.moveCursorTo(ctx, client, fx, fy);
            ctx.markEvent(.click);

            try client.dispatchMouseEvent("mousePressed", fx, fy, .{
                .button = "left",
                .click_count = 1,
            });

            actions.sleepMs(150);

            const dist = @sqrt((tx - fx) * (tx - fx) + (ty - fy) * (ty - fy));
            const timing = cursor_motion.computeDragTiming(dist);
            const waypoints = try cursor_motion.computeEasedPath(ctx.allocator, fx, fy, tx, ty, timing.steps);
            defer ctx.allocator.free(waypoints);

            for (waypoints) |wp| {
                try client.dispatchMouseEvent("mouseMoved", wp.x, wp.y, .{
                    .button = "left",
                    .buttons = 1,
                });
                if (ctx.isRecording()) {
                    if (ctx.timeline) |tl| {
                        const path_arr = [_]cursor_motion.Point{wp};
                        tl.setCursorPath(&path_arr);
                    }
                }
                actions.sleepMsF(timing.delay_ms + (prng() - 0.5) * 8.0);
            }

            try client.dispatchMouseEvent("mouseReleased", tx, ty, .{
                .button = "left",
                .click_count = 1,
            });

            ctx.setCursorPosition(tx, ty);
        },
        .moveTo => {
            const target = StepTarget{
                .text = step.text,
                .selector = step.selector,
                .within = step.within,
            };
            const box = try resolveTarget(client, target, step.timeout orelse 10000);
            const pt = actions.randomPointInBox(box, 0.1);
            try actions.moveCursorTo(ctx, client, pt.x, pt.y);
        },
        .screenshot => {
            const output = step.output orelse return error.MissingOutput;
            try actions.captureScreenshotToFile(client, output);
        },
        .navigate => {
            const url = step.url orelse return error.MissingUrl;
            try actions.navigateTo(client, url);
        },
        .hover => {
            const target = StepTarget{
                .text = step.text,
                .selector = step.selector,
                .within = step.within,
            };
            const box = try resolveTarget(client, target, step.timeout orelse 10000);
            const pt = actions.randomPointInBox(box, 0.1);
            try actions.moveCursorTo(ctx, client, pt.x, pt.y);
            try client.dispatchMouseEvent("mouseMoved", pt.x, pt.y, .{});
        },
        .select => {
            const sel = step.selector orelse return error.MissingSelector;
            const val = step.value orelse return error.MissingValue;
            try actions.selectValue(client, sel, val);
        },
    }

    if (step.delay) |d| {
        actions.sleepMs(d);
    }
}

pub fn executeSteps(ctx: *actions.RecordingContext, client: *cdp.CdpClient, steps: []const Step, default_delay: ?u64) !void {
    for (steps, 0..) |step, i| {
        executeStep(ctx, client, step) catch |err| {
            std.debug.print("Step {d} ({s}) failed: {}\n", .{ i, @tagName(step.action), err });
            return err;
        };

        if (step.delay == null) {
            if (default_delay) |d| {
                if (d > 0) actions.sleepMs(d);
            }
        }
    }
}

var runner_rng_state: u64 = 0;
var runner_rng_initialized: bool = false;

fn prng() f64 {
    if (!runner_rng_initialized) {
        runner_rng_state = @truncate(@as(u128, @bitCast(std.time.nanoTimestamp())) ^ 0xabcdef0123456789);
        runner_rng_initialized = true;
    }
    runner_rng_state ^= runner_rng_state << 13;
    runner_rng_state ^= runner_rng_state >> 7;
    runner_rng_state ^= runner_rng_state << 17;
    const float_bits = (runner_rng_state >> 12) | (0x3FF << 52);
    return @as(f64, @bitCast(float_bits)) - 1.0;
}
